### PR TITLE
Support different provisioners and Kernel clients with Kernel Gateway

### PR DESCRIPTION
**Pull request by vincentye38**
# Extensible Kernel Provisioning: Registry System and Connection Info Pattern

## Overview

This PR introduces two complementary features that make kernel provisioning extensible and maintainable:

1. **Kernel Client Registry** - Dynamic mapping system for provisioner-to-client pairing
2. **Connection Info Pattern** - Standardized communication of connection details

Together, these enable external packages to provide custom kernel clients for their provisioners without modifying core code, while maintaining a clean and consistent architecture.

---

### Problem Statement

Kernel client selection was previously configured statically in configuration:
- Fixed configuration in `c.KernelManager.client_class=`, `c.ServerApp.kernel_manager_class=`
- Provisioner-specific kernel managers manually set `client_class`
- No extensibility for external packages
- Difficult to maintain as new provisioners are added

Example of old hardcoded approach:
```python
class SparkProvisionerAwareKernelManager(KernelManager):
    client_class = DocumentAwareSparkProvisionerAwareKernelClient
    client_factory = DocumentAwareSparkProvisionerAwareKernelClient
```

### Solution: KernelClientRegistry

A singleton registry that dynamically maps provisioner types to kernel client classes.

#### Core Components

**1. KernelClientRegistry (Singleton)**
- Location: `nextgen_kernels_api/services/kernels/kernel_client_registry.py`
- Singleton using `SingletonConfigurable` pattern
- Automatic entry point discovery in `__init__()`
- Configurable fallback client (defaults to `JupyterServerKernelClient`)
- Smart lookup with exact and inheritance-based matching

**2. ProvisionerAwareKernelManager**
- Location: `nextgen_kernels_api/services/kernels/kernelmanager.py`
- Overrides `select_client()` hook to query registry
- Automatically selects appropriate client based on provisioner type
- Backward compatible with standard kernel managers

**3. Extensible KernelManager Base**
- Added `select_client()` hook method for clean separation of concerns
- Subclasses can override to customize client selection

#### Usage for Extension Developers

**Register via Entry Point (Recommended)**

In your package's `pyproject.toml`:
```toml
[project.entry-points.jupyter_kernel_client_registry]
"jupyter_server.saturn.provisioners.spark_provisioner:SparkProvisioner" = 
    "jupyter_server_documents.kernel_client:DocumentAwareSparkProvisionerAwareKernelClient"
```

The registry automatically discovers and loads these on initialization!

**Register Programmatically**
```python
from nextgen_kernels_api.services.kernels.kernel_client_registry import KernelClientRegistry

KernelClientRegistry.register(MyProvisioner, MyKernelClient)
```

#### Configuration for End Users

In `jupyter_config.py`:
```python
# Set custom fallback client
c.KernelClientRegistry.fallback_client_class = "jupyter_server_documents.kernel_client.DocumentAwareKernelClient"

# Use provisioner-aware kernel manager
c.MultiKernelManager.kernel_manager_class = "nextgen_kernels_api.services.kernels.kernelmanager.ProvisionerAwareKernelManager"
```

#### Registry Benefits

✅ **Extensibility** - External packages register via entry points  
✅ **Flexibility** - Multiple registration methods supported  
✅ **Backward Compatible** - Existing code continues to work  
✅ **Clean Separation** - Registry handles mapping logic  
✅ **Smart Lookup** - Exact → inheritance → fallback  

#### Files Changed (Registry)

**New Files:**
- `nextgen_kernels_api/services/kernels/kernel_client_registry.py` - Registry implementation
- `tests/test_kernel_client_registry.py` - 33 comprehensive unit tests
- `docs/provisioner_client_registry_design.md` - Architecture documentation

**Modified Files:**
- `nextgen_kernels_api/services/kernels/kernelmanager.py` - Added `select_client()` hook
- `jupyter_config.py` - Configuration examples
- `pyproject.toml` - Example entry point

---

### Problem Statement

Different provisioner types handled connection information inconsistently:

**LocalProvisioner** (ZMQ-based):
- Sets ports on KernelManager attributes: `km.shell_port = 5555` ✓
- Stores connection_info for retrieval ✓

**SparkProvisioner** (Gateway-based):
- Generates WebSocket URL internally
- Returns ZMQ ports in `connection_info` (incorrect!)
- GatewayKernelManager had to generate WebSocket URL itself

This created issues:
- Tight coupling: GatewayKernelManager needed provisioner-specific logic
- Not extensible: New provisioners required modifying KernelManager
- Inconsistent: Different provisioners followed different patterns
- Confusing: SparkProvisioner returned wrong connection type

### Solution: Standardized Connection Info Delegation

Implement a consistent pattern where:

1. **Provisioners are the source of truth** for connection details
2. **`connection_info` dictionary** carries ANY fields the provisioner needs
3. **KernelManager delegates** to `provisioner.connection_info`
4. **Clients interpret** connection_info based on their type

#### Architecture

```
Provisioner stores          KernelManager delegates      Client loads
connection details    →     to provisioner          →    and interprets

LocalProvisioner            KernelManager                JupyterServerKernelClient
  {shell_port: 5555,   →    get_connection_info()   →    load_connection_info()
   iopub_port: 5556,         provisioner.connection_info  uses ZMQ ports
   ...}                      

SparkProvisioner            KernelManager                GatewayKernelClient
  {ws_url: "ws://...",  →   get_connection_info()   →    load_connection_info()
   key: b"..."}              provisioner.connection_info  uses WebSocket URL
```

#### Changes Made (Connection Info)

**1. KernelManager - Generic Delegation**

`nextgen_kernels_api/services/kernels/kernelmanager.py`:
```python
def get_connection_info(self, session: bool = False) -> dict:
    """Get connection info by delegating to provisioner."""
    if self.provisioner and hasattr(self.provisioner, 'connection_info') and self.provisioner.connection_info:
        # NEW: Delegate to provisioner (extensible pattern)
        info = self.provisioner.connection_info.copy()
    else:
        # Fallback: Build from KM attributes (backward compatibility)
        info = {...}
    
    if session:
        info["key"] = self.session.key
    
    return info
```

**2. SparkProvisioner - Store WebSocket Info**

`jupyter_server/saturn/provisioners/spark_provisioner.py`:
```python
def __init__(self, **kwargs):
    super().__init__(**kwargs)
    self._connection_info = {}  # NEW

async def submit_spark_job(self, **kwargs):
    # ... launch Spark job ...
    
    # NEW: Store WebSocket connection info
    self._connection_info = {
        "ws_url": self.jupyter_gateway_url,
        "key": km.session.key,
        "signature_scheme": "hmac-sha256",
        "kernel_id": self.kernel_id,
        "submission_id": self.submission_id,
    }

@property
def connection_info(self):
    return self._connection_info  # CHANGED: Was returning ZMQ ports
```

**3. GatewayKernelClient - Expect WebSocket URL**

`nextgen_kernels_api/gateway/managers.py`:
```python
def load_connection_info(self, info):
    """Load WebSocket connection info from provisioner."""
    # NEW: Validate ws_url present
    if "ws_url" not in info:
        raise ValueError(
            "GatewayKernelClient requires 'ws_url' in connection_info. "
            "This should be provided by the provisioner."
        )
    
    self.ws_url = info["ws_url"]
    # Load session key if present
    if "key" in info:
        self.session.key = info["key"]
```

#### Connection Info Benefits

✅ **Extensibility** - Add new provisioner types without changing KernelManager  
✅ **Generic KernelManager** - No provisioner-specific attributes needed  
✅ **Type Safety** - Registry ensures compatible provisioner-client pairs  
✅ **Consistency** - All provisioners follow same contract  
✅ **Backward Compatible** - Existing LocalProvisioner pattern preserved  

#### Future Provisioner Examples

```python
# Kubernetes provisioner
self._connection_info = {
    "k8s_pod_name": "jupyter-pod-123",
    "k8s_service_url": "http://...",
    "key": session_key
}

# SSH provisioner  
self._connection_info = {
    "ssh_host": "remote-server.com",
    "ssh_port": 22,
    "kernel_ports": {...},
    "key": session_key
}
```

---

### Registry System Tests
- **33 unit tests** across 9 test classes
- All tests passing ✅
- Coverage includes:
  - Singleton behavior
  - Registration methods (programmatic, string-based)
  - Client lookup (exact, inheritance, fallback)
  - Entry point auto-discovery
  - Configuration via strings (tested and working)
  - Real entry point validation from pyproject.toml
  - Edge cases and error handling

### Run Tests
```bash
cd nextgen-kernels-api
pip install -e ".[dev]"
pytest tests/test_kernel_client_registry.py -v
```

### Manual Testing for Connection Info
```python
# Test 1: LocalProvisioner still works
provisioner = LocalProvisioner()
await provisioner.launch_kernel(cmd=["python", "-m", "ipykernel"])
assert "shell_port" in provisioner.connection_info

# Test 2: SparkProvisioner uses WebSocket
provisioner = SparkProvisioner()
await provisioner.submit_spark_job()
assert "ws_url" in provisioner.connection_info
assert "ws://" in provisioner.connection_info["ws_url"]

# Test 3: Registry mapping works
client_class = KernelClientRegistry.instance().get_client_for_provisioner(provisioner)
assert client_class is GatewayKernelClient

# Test 4: ProvisionerAwareKernelManager
manager = ProvisionerAwareKernelManager()
await manager.start_kernel()
assert manager.client_class == GatewayKernelClient
```

---

#### Before (Hardcoded)
```python
class SparkProvisionerAwareKernelManager(KernelManager):
    client_class = DocumentAwareSparkProvisionerAwareKernelClient
    client_factory = DocumentAwareSparkProvisionerAwareKernelClient
```

#### After (Registry-Based)

**In pyproject.toml:**
```toml
[project.entry-points.jupyter_kernel_client_registry]
"jupyter_server.saturn.provisioners.spark_provisioner:SparkProvisioner" = 
    "jupyter_server_documents.kernel_client:DocumentAwareSparkProvisionerAwareKernelClient"
```

**Use generic manager:**
```python
c.MultiKernelManager.kernel_manager_class = "nextgen_kernels_api.services.kernels.kernelmanager.ProvisionerAwareKernelManager"
```

### For Custom Provisioner Developers

Follow the connection info pattern:

```python
class MyProvisioner(KernelProvisionerBase):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._connection_info = {}  # Store connection details here
    
    async def launch_kernel(self, cmd, **kwargs):
        # ... launch kernel ...
        
        # Store connection details with fields your client expects
        self._connection_info = {
            "my_custom_field": value,
            "key": self.parent.session.key,
        }
        return self.connection_info
    
    @property
    def connection_info(self):
        return self._connection_info  # Expose via property
```

Then register your client via entry point or programmatically.

---

### ⚠️ Potential Breaking Changes

1. **SparkProvisioner.connection_info format changed**
   - **Before**: Returned ZMQ ports (incorrect)
   - **After**: Returns WebSocket URL (correct)
   - **Impact**: Code directly accessing `SparkProvisioner.connection_info` expecting ZMQ ports
   - **Mitigation**: This was already broken - SparkProvisioner never used ZMQ ports

2. **GatewayKernelClient.load_connection_info() now validates ws_url**
   - **Before**: No validation
   - **After**: Raises ValueError if `ws_url` not in connection_info
   - **Impact**: Improves error messages for misconfigured setups
   - **Mitigation**: Clear error message guides users to fix configuration

**Note:** For most users, no changes needed. Existing code continues to work.

---

### New Documentation Files

1. **`docs/connection_info_pattern.md`** - Connection info architecture and patterns
2. **`docs/provisioner_client_registry_design.md`** - Registry architecture and API reference  
3. **`docs/provisioner_developer_guide.md`** - Complete developer guide with examples

### Updated Configuration

**`jupyter_config.py`** - Added examples for:
- Configuring fallback client
- Using ProvisionerAwareKernelManager
- Entry point registration format

---

### New Files

1. **Registry Implementation**
   - `nextgen_kernels_api/services/kernels/kernel_client_registry.py` - Registry singleton

2. **Tests**
   - `tests/test_kernel_client_registry.py` - 33 comprehensive unit tests

3. **Documentation**
   - `docs/provisioner_client_registry_design.md` - Registry architecture
   - `docs/connection_info_pattern.md` - Connection info patterns
   - `docs/provisioner_developer_guide.md` - Developer guide

### Modified Files

1. **Kernel Manager**
   - `nextgen_kernels_api/services/kernels/kernelmanager.py`
     - Added `select_client()` hook to base `KernelManager`
     - Implemented `ProvisionerAwareKernelManager`
     - Added connection info delegation in `get_connection_info()`

2. **Gateway Components** 
   - `nextgen_kernels_api/gateway/managers.py`
     - Updated `GatewayKernelClient.load_connection_info()` validation
     - Removed WebSocket URL generation from `GatewayKernelManager`

3. **Provisioners**
   - `jupyter_server/saturn/provisioners/spark_provisioner.py`
     - Store WebSocket connection info in `_connection_info`
     - Return correct connection details via `connection_info` property

4. **Configuration**
   - `jupyter_config.py` - Added configuration examples
   - `pyproject.toml` - Added example entry point

---

### 1. Extensibility
- External packages register via entry points
- No core code modification needed
- Easy to add new provisioner types

### 2. Flexibility
Three registration methods:
- Entry points (automatic discovery)
- Programmatic registration
- Configuration-based fallback

### 3. Clean Separation
- Provisioners store connection details
- KernelManager delegates without knowing specifics
- Registry handles type mapping
- Clients interpret their expected format

### 4. Generic Components
- **KernelManager**: No provisioner-specific attributes
- **ProvisionerAwareKernelManager**: Works with any provisioner type
- **Registry**: Handles all mapping logic

### 5. Type Safety
Registry ensures provisioners pair with compatible clients:
```python
LocalProvisioner → JupyterServerKernelClient (ZMQ ports)
SparkProvisioner → GatewayKernelClient (WebSocket URL)
FutureProvisioner → CustomClient (custom fields)
```

### 6. Consistency
All provisioners follow the same contract:
1. Store connection details in `_connection_info`
2. Expose via `@property connection_info`
3. KernelManager retrieves and passes to client

---

### Automated Tests
- ✅ **33 unit tests** for kernel client registry
- ✅ All tests passing
- ✅ Singleton behavior validated
- ✅ Entry point discovery tested
- ✅ String configuration tested
- ✅ Real pyproject.toml entry point validated
- ✅ Inheritance matching tested
- ✅ Edge cases covered

### Manual Testing Scenarios
```python
# Test 1: LocalProvisioner (ZMQ)
provisioner = LocalProvisioner()
await provisioner.launch_kernel()
assert "shell_port" in provisioner.connection_info

# Test 2: SparkProvisioner (WebSocket)
provisioner = SparkProvisioner()
await provisioner.submit_spark_job()
assert "ws_url" in provisioner.connection_info

# Test 3: Registry selection
registry = KernelClientRegistry.instance()
client_class = registry.get_client_for_provisioner(provisioner)
assert client_class == GatewayKernelClient

# Test 4: ProvisionerAwareKernelManager
manager = ProvisionerAwareKernelManager()
await manager.start_kernel()
assert manager.client_class == GatewayKernelClient
```

---

### Entry Point Registration
```toml
[project.entry-points.jupyter_kernel_client_registry]
"your.package.provisioner:YourProvisioner" = "your.package.client:YourKernelClient"
```

### Fallback Client Configuration
```python
c.KernelClientRegistry.fallback_client_class = "your.package.YourDefaultClient"
```

### Using ProvisionerAwareKernelManager
```python
c.ServerApp.kernel_manager_class = "nextgen_kernels_api.services.kernels.kernelmanager.MultiKernelManager"
c.MultiKernelManager.kernel_manager_class = "nextgen_kernels_api.services.kernels.kernelmanager.ProvisionerAwareKernelManager"
```

---

## Backward Compatibility

✅ **No Breaking Changes for Standard Usage**
- Existing kernel managers continue to work
- LocalProvisioner unchanged (already followed pattern)
- Graceful fallback when `connection_info` not available
- No changes required for users

⚠️ **Potential Impact for Custom Code**
- Code directly accessing `SparkProvisioner.connection_info` expecting ZMQ ports (was already broken)
- Subclasses overriding `GatewayKernelManager.get_connection_info()` (rare)

---

### Implementation
- ✅ Kernel Client Registry implemented
- ✅ Entry point auto-discovery working
- ✅ Connection info delegation pattern implemented
- ✅ `select_client()` hook added to base KernelManager
- ✅ ProvisionerAwareKernelManager complete
- ✅ SparkProvisioner updated with correct connection info
- ✅ GatewayKernelClient validation added

### Testing
- ✅ 33 unit tests written and passing
- ✅ String configuration tested
- ✅ Real entry point validated
- ✅ Inheritance matching tested
- ⏳ Integration tests with real Spark clusters (manual testing)

### Documentation
- ✅ API documentation complete
- ✅ Architecture documentation added
- ✅ Developer guide with examples
- ✅ Configuration examples provided
- ✅ Migration guide included

### Quality
- ✅ Backward compatibility maintained
- ✅ No breaking changes for standard usage
- ✅ Clear error messages added
- ✅ Comprehensive docstrings

---

## Related

This PR enables:
- External packages to provide custom kernel clients for their provisioners
- Spark provisioner to use document-aware clients without hardcoding
- Future provisioners (Docker, Kubernetes, SSH) to be easily added
- Configuration-driven client selection for deployment flexibility
- Clean architecture with separation of concerns

## Reviewers

Please pay special attention to:
1. Registry singleton pattern and configuration handling
2. Connection info delegation logic
3. SparkProvisioner WebSocket URL handling
4. Backward compatibility with LocalProvisioner
5. Test coverage and edge cases
6. Documentation clarity

### CODE REVIEW

File: docs/connection_info_pattern.md [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+### Issue: Client doesn't understand connection_info fields
+
+**Solution**: Ensure provisioner and client are paired correctly in registry and client's `load_connection_info()` method handles the expected fields.
+
+## Related Files
+
+- [`nextgen_kernels_api/services/kernels/kernelmanager.py`](../nextgen_kernels_api/services/kernels/kernelmanager.py) - KernelManager with delegation
+- [`nextgen_kernels_api/services/kernels/kernel_client_registry.py`](../nextgen_kernels_api/services/kernels/kernel_client_registry.py) - Registry system
+- [`nextgen_kernels_api/gateway/managers.py`](../nextgen_kernels_api/gateway/managers.py) - GatewayKernelClient
+- `jupyter_server/saturn/provisioners/spark_provisioner.py` - SparkProvisioner implementation
```
Comment (vincentye38) [inline docs/connection_info_pattern.md:345]: remove saturn

File: nextgen_kernels_api/services/kernels/kernelmanager.py [hunk 2]
```diff
@@ -43,9 +46,6 @@ def __init__(self, **kwargs):
         """Initialize the kernel manager and create a kernel client instance."""
         super().__init__(**kwargs)
 
-        # Create a kernel client instance immediately
-        self.kernel_client = self.client(session=self.session)
-
```
Comment (vincentye38) [inline nextgen_kernels_api/services/kernels/kernelmanager.py:48]: Moved this block to _async_post_start_kernel()
